[OnWeek][Discover] Allow to fetch more documents on Discover page

jughosta opened the pull request:
- Closes https://github.com/elastic/kibana/issues/155019

Per docs https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html
<img width="851" alt="Screenshot 2023-05-10 at 10 25 20" src="https://github.com/elastic/kibana/assets/1415710/b4b9fef4-7dd8-40ed-8244-343889fc4367">


## Summary

1. This PR improves `search_after` pagination for `date_nanos` time fields. `sort` value will be returned from ES as a string instead of a rounded and incorrect timestamp. This change allows to also simplify logic on Surrounding document page.

Before:
<img width="400" alt="Screenshot 2023-05-08 at 17 36 19" src="https://github.com/elastic/kibana/assets/1415710/fd9f45c4-5dc2-4103-83b9-8810e3a6e0df">

After:
<img width="400" alt="Screenshot 2023-05-08 at 17 37 13" src="https://github.com/elastic/kibana/assets/1415710/fe9090c0-2116-4f77-9a57-a96ae6b00365">

2. Also in this PR we now allow users to load more documents within the same time range. Once the button is pressed, it will load next portion of documents (same "sampleSize" value will be used). Currently, we limit max total loaded documents to 10000.

"Load more" demo:
![Aug-07-2023 16-23-28](https://github.com/elastic/kibana/assets/1415710/53af9809-75cb-4b8a-8e99-d8f6d76b4981)

If refresh interval is on, the button becomes disabled:
![Aug-07-2023 16-24-58](https://github.com/elastic/kibana/assets/1415710/85db6144-98eb-40b5-ac88-80ea728bcd6b)

Date nanos demo:
![Aug-07-2023 16-34-59](https://github.com/elastic/kibana/assets/1415710/dc9fe0b1-e419-4c76-9fc6-79907b134e58)


100x Flaky test runner https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/2801

jughosta committed:
[Discover] Improve date_nanos support for "search_after" pagination

jughosta committed:
Merge branch 'main' into onweek-date-nanos-pagination

jughosta committed:
[Discover] Allow to fetch more documents on Discover page

jughosta committed:
Merge remote-tracking branch 'origin/onweek-date-nanos-pagination' into onweek-date-nanos-pagination

jughosta committed:
[Discover] Fix when the footer should appear

jughosta committed:
[Discover] Experiment with a tie breaker field

jughosta committed:
[Discover] Update other places

jughosta committed:
[Discover] Comment out the tie breaker

jughosta committed:
[Discover] Fix lint issues

kertal commented on src/plugins/discover/common/utils/sorting/get_sort_for_search_source.ts:
```diff
@@ -31,16 +40,31 @@ export function getSortForSearchSource(
     return sortPair as EsQuerySortValue;
   });
+
+  // TODO: do we need to have a tie breaker like this?
```
Would it be a tie breaker like mentioned in this Issue ? https://github.com/elastic/kibana/issues/47934

jughosta replied:
@kertal Yes, in this implementation I am using a tie breaker as it's done in context app: based on Advanced Setting `context:tieBreakerFields`. I think it makes sense to add it but this change can be considered as a breaking change (also for CSV). So I would need your advice on whether to add it now or not.

kertal replied:
Good point about the breaking change. Is it required or optional for the implementation? In terms of breaking change, it don't think it would break existing functionality inside Discover, but, yes, mainly it's about CSV, it could lead to a different sort order when exporting and the tie breaker is not implemented.

Would it be an option to exclude the tiebreaker for this PR and implement it in a follow up / later, with also CSV included?

jughosta replied:
@kertal Current implementation of CSV relies on Discover search source `sort` param and reuses it's value. That means that if a tie breaker is not used for Discover then it's not used for CSV generation at the moment. @tsullivan please correct if my assumption is wrong. It's a good question if an explicit tie breaker is required or not.

kertal replied:
just to make sure, since we continued the discussion "offline" 
@jughosta mentioned
> CSV export is done with search_after and pit so I guess according to docs it is running with _shard_doc by default

So if we introduce a tiebreaker for Discover like `_doc`  that we use for incremental data fetching, this would not influence CSV generation, on the contrary, it would make the data fetching more similar, since we'd also use a tie breaker. Can you confirm @tsullivan @rshen91 ? Many thx!

jughosta replied:
Quick update here: added `_doc` as a tie breaker for Discover only so far. It would not affect CSV at this point.

kertal commented on src/plugins/discover/public/application/main/utils/fetch_all.ts:
```diff
@@ -131,6 +132,70 @@ export function fetchAll(
+    const query = getAppState().query;
+    const recordRawType = getRawRecordType(query);
+
+    if (recordRawType === RecordRawType.PLAIN) {
+      // not supported yet
+      // TODO?
+      return Promise.resolve();
+    }
```
I don't think for the initial version this is required, also I assume the logic would be very different, right?

kertal commented on src/plugins/discover/public/application/main/utils/fetch_all.ts:
```diff
@@ -131,6 +132,70 @@ export function fetchAll(
+          query,
+        });
+      })
+      .catch((error) => {
+        // TODO?
+        // eslint-disable-next-line no-console
+        console.error(error);
+      });
```
Ideally yes, some sort of error indication would be great, currently users just get an infinite spinning link when it doesn't work

kertal replied:
tested, and there's an error toast, thx and resolving conversation

kertal commented on src/plugins/discover/server/locator/searchsource_from_locator.ts:
```diff
@@ -147,7 +147,7 @@ export function searchSourceFromLocatorFactory(services: LocatorServicesDeps) {
     // Inject sort
     if (savedSearch.sort) {
-      const sort = getSortForSearchSource(savedSearch.sort as Array<[string, string]>, index);
+      const sort = getSortForSearchSource(savedSearch.sort as Array<[string, string]>, index, null); // TODO: is it correct that it was using a `desc` as default sort here?
```
No, I think now it's correct, thx for catching!

kertal commented on src/plugins/discover/public/components/discover_grid/discover_grid_footer.tsx:
```diff
@@ -0,0 +1,111 @@
+
+  // allow to fetch more records on Discover page
+  if (onFetchMoreRecords) {
+    // TODO: define a limit
+    if (rowCount <= 3000 - sampleSize) {
+      return (
+        <DiscoverGridFooterContainer hasButton={true} {...props}>
```
Finding the right limit is a good question, it very much depends on the size of the records how much browsers resources / memory  fetching more data would take. dear @weltenwort, having more experience with incremental loading, what's the current limit of Logs, you recently mentioned it to me, thx!
Don't think we need to make this setting configurable until requested , but we should provide it in :
https://github.com/elastic/kibana/blob/92a46f53444c6e419967afa55b2e7664b7c9615a/src/plugins/discover/public/components/discover_grid/constants.ts#L14

kertal replied:
thx @weltenwort , resolving comment since it's old, and outdated

kertal reviewed (commented):
This will be a great little improvement, breaking out of the usual 500 limit is very useful 🥳 
One thing that that would be neat, currently, when loading more data, also histogram data is loaded. But in this case, just loading the document should be sufficient.

kertal commented:
Adding some notes of our planning:
* We need to figure out if the maximum of 3000 is a reasonable number (when testing)
* In case auto-refresh is on, the load more link should be disabled (with a tooltip to inform the user)

jughosta committed:
Merge remote-tracking branch 'upstream/main' into onweek-date-nanos-pagination

# Conflicts:
#	src/plugins/discover/public/application/context/hooks/use_context_app_fetch.tsx
#	src/plugins/discover/public/application/context/services/context.ts
#	src/plugins/discover/public/application/main/components/layout/discover_documents.tsx
#	src/plugins/discover/public/application/main/services/discover_data_state_container.ts
#	src/plugins/discover/public/components/discover_grid/discover_grid.tsx

jughosta committed:
[Discover] Update styles

jughosta committed:
[Discover] Handle error cases

jughosta committed:
[Discover] Fix types

jughosta committed:
[Discover] Fix default sorting order for locator

jughosta committed:
[Discover] Add a comment

jughosta committed:
[Discover] Add a tooltip when refresh interval is on

jughosta committed:
[Discover] Update code style

jughosta committed:
Merge branch 'main' into onweek-date-nanos-pagination

jughosta committed:
[Discover] Show error toasts

jughosta committed:
Merge remote-tracking branch 'upstream/main' into onweek-date-nanos-pagination

# Conflicts:
#	src/plugins/discover/public/application/context/hooks/use_context_app_fetch.tsx
#	src/plugins/discover/public/application/main/utils/update_search_source.ts
#	src/plugins/discover/public/utils/get_sharing_data.ts

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

jughosta committed:
Merge remote-tracking branch 'upstream/main' into onweek-date-nanos-pagination

jughosta committed:
[Discover] Fix after merging main

jughosta committed:
Merge remote-tracking branch 'upstream/main' into onweek-date-nanos-pagination

jughosta committed:
Merge remote-tracking branch 'upstream/main' into onweek-date-nanos-pagination

jughosta committed:
[Discover] Update comment

jughosta committed:
[Discover] Remove redundant param

jughosta committed:
[Discover] Refactor params of getSortForSearchSource

jughosta committed:
[Discover] Refactor params of get_es_query_sort

jughosta committed:
[Discover] Don't update histogram for "Load more" in the grid

jughosta committed:
[Discover] Fix footer for Surrounding Documents and Dashboard pages

jughosta committed:
[Discover] Add unit tests

jughosta committed:
[Discover] Add tests for the footer

jughosta committed:
[Discover] Add functional tests

jughosta committed:
Merge branch 'main' into onweek-date-nanos-pagination

jughosta committed:
[Discover] Add functional tests for date nanos

jughosta committed:
Merge remote-tracking branch 'origin/onweek-date-nanos-pagination' into onweek-date-nanos-pagination

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

jughosta committed:
[Discover] Extract the code

jughosta committed:
Merge remote-tracking branch 'origin/onweek-date-nanos-pagination' into onweek-date-nanos-pagination

jughosta committed:
Merge branch 'main' into onweek-date-nanos-pagination

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

jughosta committed:
Merge branch 'main' into onweek-date-nanos-pagination

andreadelrio reviewed (approved):
LGTM, thanks!

andreadelrio commented on src/plugins/discover/public/components/discover_grid/discover_grid.scss:
```diff
@@ -65,14 +65,6 @@
-  flex-shrink: 0;
-  background-color: $euiColorLightShade;
-  padding: $euiSize / 2 $euiSize;
-  margin-top: $euiSize / 4;
```
Here you can use more specific EUI sizing variables, for example `$euiSizeS`, `$euiSizeXS` that way you don't need the divisions.

jughosta replied:
@andreadelrio This code is removed now. In `src/plugins/discover/public/components/discover_grid/discover_grid_footer.tsx` I use emotion instead.

weltenwort commented on src/plugins/discover/public/application/main/components/layout/use_discover_histogram.ts:
```diff
@@ -279,7 +279,10 @@ export const useDiscoverHistogram = ({
     } else {
-      fetch$ = stateContainer.dataState.fetch$.pipe(map(() => 'discover'));
+      fetch$ = stateContainer.dataState.fetch$.pipe(
+        filter(({ options }) => !options.fetchMore), // don't update histogram for "Load more" in the grid
```
Do any of the histogram unit tests cover this case?

jughosta replied:
@weltenwort Added via https://github.com/elastic/kibana/pull/157241/commits/a57eafa030f56deafba0cd1519111ef31f53fee8, thanks!

weltenwort commented on src/plugins/discover/public/application/main/services/discover_data_state_container.ts:
```diff
@@ -219,6 +224,27 @@ export function getDataStateContainer({
+      };
+
+      if (options.fetchMore) {
+        abortControllerFetchMore?.abort();
```
Wouldn't we want to cancel both kinds of requests regardless of the case?

jughosta replied:
@weltenwort Right! Updated via https://github.com/elastic/kibana/pull/157241/commits/5d10203af084e0b4542704bef6124c4d6acb6796

weltenwort commented on src/plugins/discover/public/components/discover_grid/discover_grid.tsx:
```diff
@@ -93,6 +94,10 @@ export interface DiscoverGridProps {
+  /**
+   * Determines if it's loading more data (next chunk)
+   */
+  isLoadingMore?: boolean;
```
Not a request to change anything here, but a thought on data structure design :thinking: since it can't be `loading` and `loading_more` at the same time, a single field which indicates either `loading`, `loading_more` or `idle` could represent this in a way that makes invalid state unrepresentable.

weltenwort reviewed (commented):
Thanks for notifying me about this enhancement. I'm excited it's actually happening. It's really impressive that you could make it work given the complexity of the app.

Since other reviewers are better suited to comment on the changes to Discover as such I focused mainly on the data fetching itself.

jughosta committed:
Merge remote-tracking branch 'upstream/main' into onweek-date-nanos-pagination

# Conflicts:
#	src/plugins/discover/public/application/context/hooks/use_context_app_fetch.tsx
#	src/plugins/discover/public/application/context/services/context.ts
#	src/plugins/discover/public/application/main/utils/fetch_documents.ts
#	src/plugins/discover/public/embeddable/saved_search_grid.tsx
#	src/plugins/discover/tsconfig.json

jughosta commented on src/plugins/discover/common/utils/sorting/get_sort_for_search_source.ts:
```diff
@@ -31,16 +40,34 @@ export function getSortForSearchSource(
     return sortPair as EsQuerySortValue;
   });
+
+  // Do we need to have a tie breaker like this? We have it for Surrounding Documents page but not here yet.
```
@weltenwort Good point, thanks! Would you suggest to reuse `context:tieBreakerFields` setting or have a fixed tie breaker like `_shard_doc`?

weltenwort replied:
If I understand the docs correctly `_shard_doc` is only available within PIT contexts, although it would be a great option. But maybe I'm mistaken there. Otherwise it seems reasonable to me to re-use the existing tiebreaker setting.

jughosta committed:
[Discover] Add a test for fetchMore

jughosta committed:
[Discover] Changes to abort logic

jughosta commented on src/plugins/discover/common/utils/sorting/get_es_query_sort.ts:
```diff
@@ -0,0 +1,102 @@
+ * The list of field names that are allowed for sorting, but not included in
+ * data view fields.
+ */
+const META_FIELD_NAMES: string[] = ['_seq_no', '_doc', '_uid'];
```
@weltenwort Thanks for bring our attention to this! I have not introduced this array, just moved from context utils to common utils. For supporting legacy behaviour, I think we should keep `_uid` available for user's selection. We can also add `_shard_doc` to the list, wdyt?

weltenwort replied:
I think we can only use `_shard_doc` within a PIT context.

kertal replied:
```
The _shard_doc value is the combination of the shard index within the PIT and the Lucene’s internal doc ID, it is unique per document and constant within a PIT.
```
wouldn't `_doc` be sufficient, since we are also using it already in Discover:
https://github.com/elastic/kibana/blob/807625c108a6f9f67875cd891023e8e8a4926ee7/src/plugins/discover/common/utils/sorting/get_sort_for_search_source.ts#L29

weltenwort replied:
It would work, it might just lead to duplicate rows sometimes when shard merges occur. For the Logs UI that's a trade-off that we accepted.

kertal replied:
I think this trade-off is fine

jughosta committed:
Merge remote-tracking branch 'upstream/main' into onweek-date-nanos-pagination

jughosta committed:
[Discover] Add `_doc` as a tie breaker

jughosta committed:
[Discover] Update tests

kertal commented on packages/kbn-generate-csv/src/generate_csv.ts:
```diff
@@ -356,6 +356,9 @@ export class CsvGenerator {
         // set the latest pit, which could be different from the last request
         searchSource.setField('pit', { id: pitId, keep_alive: settings.scroll.duration });
 
+        // TODO: should we skip `_doc` (the tie breaker from Discover) from searchSource `sort` field when generating CSV?
+        // According to docs, the default `_shard_doc` is a better option when PIT is used.
+
         const results = await this.doSearch(searchSource, settings, searchAfter);
```
@jughosta QQ: we don't store the `_doc` field in out searchSource, we just use `getSortForSearchSource` for getting the searchSource we send to Elasticsearch, right? So why do we need to remove it, or am I overlooking something and _doc is added the the persisted saved search? thx!

jughosta replied:
@kertal I am applying some changes to skip it for other cases besides Discover page and embeddable.

kertal replied:
makes sense, thx!

jughosta committed:
[Discover] Fix for CSV

jughosta committed:
[Discover] Invert the parameter

jughosta committed:
[Discover] Remove the comment

jughosta committed:
Merge branch 'main' into onweek-date-nanos-pagination

PhilippeOberti committed:
[Security Solution] expandable flyout - expandable panel UI update (#163449)

nreese committed:
[controls] fix Dashboard getting stuck at loading in Kibana when Controls is used and mapping changed from integer to keyword (#163529)

Closes https://github.com/elastic/kibana/issues/162474

### Changes
* RangeSliderEmbeddable - call setInitializationFinished when
runRangeSliderQuery throws. This fixes the issue
* Investigated if OptionsListEmbeddable is vulnerable to the same issue.
It's not because it uses its own REST API that has a service wrapper
`OptionsListService`. `OptionsListService` handles REST API errors.
* Add unit test verifying OptionsListService.runOptionsListRequest does
not throw when there are REST API errors and always returns a response.
* Add unit tests ensuring setInitializationFinished is called for both
RangeSliderEmbeddable and OptionsListEmbeddable in all cases
* Other clean up
* Fix uses of `dataViewsService.get`. `dataViewsService.get` throws when
data view is not found. It does not return undefined. PR updates
OptionsListEmbeddable, RangeSliderEmbeddable, and mocked data service
* Fix uses of `dataView.getFieldByName`. `dataView.getFieldByName`
returns undefined when field is not found and never throws. PR updates
OptionsListEmbeddable and RangeSliderEmbeddable
    * Remove `resp` wrapper around mocked `fetch` results.

### Test instructions
1) In console run 
  ```
  PUT test1

  PUT test1/_mapping
  {
    "properties": {
      "value": {
        "type": "integer"
      }
    }
  }

  PUT test1/_doc/1
  {
      "value" : 1
  }

  PUT test1/_doc/2
  {
      "value" : 10
  }
  ```
2) create data view `test*`
3) create dashboard with range slider control on test*.value.
4) select a range in the range slider
5) save dashboard
6) run the following in console
  ```
  PUT test2

  PUT test2/_mapping
  {
    "properties": {
      "value": {
        "type": "keyword"
      }
    }
  }

  PUT test2/_doc/1
  {
      "value" : "foo"
  }

  DELETE test1
  ```
7) Open dashboard saved above. Verify dashboard opens and control
displays an error message about being unable to run aggregation on
keyword field.

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Devon Thomson <devon.thomson@elastic.co>

spong committed:
[Security Solution] Adds RBAC for Assistant (#163031)

## Summary

Adds `All`/`None` RBAC for the Elastic AI Assistant within the Security
app via Kibana feature privileges, and also via serverless PLI App
Features for the Security `complete` product (see
https://github.com/elastic/security-team/issues/7023).

Added as high-level category to enable future support of sub-features
(included sample `Create Conversation` sub-feature plumbed as example).

<p align="center">
<img width="500"
src="https://github.com/elastic/kibana/assets/2946766/b42ab3fe-65e1-49b9-a379-262f2438c0eb"
/>
</p> 

Note: Since [`minimumLicense:
'enterprise'`](https://github.com/elastic/kibana/pull/163031/files#diff-56de8b6234daf4e0e69efe680e5a4afc4f88d152243b773d90c3991fa9dabc19R28)
is configured on this privilege, when the license check isn't satisfied,
the privilege will be hidden (not disabled).

## Testing

Feature is available for `enterprise` licenses and when `All` privilege
is set, otherwise Assistant is hidden in Timeline, all `Chat` UI
elements are hidden, and the `cmd/ctrl + ;` shortcut is unavailable.

---

### On Prem Testing:
Create two roles, one for each `all`/`none` Security Elastic AI
Assistant privilege (via Dev Tools):

<details><summary>PUT /_security/role/assistant_all</summary>
<p>

``` ts
PUT /_security/role/assistant_all
{
  "cluster": [
    "all"
  ],
  "indices": [
    {
      "names": [
        "*"
      ],
      "privileges": [
        "all"
      ],
      "field_security": {
        "grant": [
          "*"
        ],
        "except": []
      },
      "allow_restricted_indices": false
    }
  ],
  "applications": [
    {
      "application": "kibana-.kibana",
      "privileges": [
        "feature_securitySolutionAssistant.minimal_all",
        "feature_siem.all",
        "feature_securitySolutionCases.all",
        "feature_actions.all"
      ],
      "resources": [
        "*"
      ]
    }
  ],
  "run_as": [],
  "metadata": {},
  "transient_metadata": {
    "enabled": true
  }
}
```
</p>
</details>


<details><summary>PUT /_security/role/assistant_none</summary>
<p>

``` ts
PUT /_security/role/assistant_none
{
  "cluster": [
    "all"
  ],
  "indices": [
    {
      "names": [
        "*"
      ],
      "privileges": [
        "all"
      ],
      "field_security": {
        "grant": [
          "*"
        ],
        "except": []
      },
      "allow_restricted_indices": false
    }
  ],
  "applications": [
    {
      "application": "kibana-.kibana",
      "privileges": [
        "feature_siem.all",
        "feature_securitySolutionCases.all",
        "feature_actions.all"
      ],
      "resources": [
        "*"
      ]
    }
  ],
  "run_as": [],
  "metadata": {},
  "transient_metadata": {
    "enabled": true
  }
}
```
</p>
</details>

Create a new `assistant_user` (assigned both roles above), log in and
test assistant availability, then remove one role at a time testing
each:

<details><summary>POST /_security/user/assistant_user (ALL)</summary>
<p>

``` ts
POST /_security/user/assistant_user
{
    "username": "assistant_user",
    "password": "changeme",
    "roles": [
      "assistant_all",
      "assistant_none"
    ],
    "full_name": "Assistant User",
    "email": "assistant-user@elastic.co",
    "metadata": {},
    "enabled": true
}
```
</p>
</details>

Test that assistant is available in UI via `Chat` buttons and shortcut
keys.

<details><summary>PUT /_security/user/assistant_user (NONE)</summary>
<p>

``` ts
PUT /_security/user/assistant_user
{
    "username": "assistant_user",
    "roles": [
      "assistant_none"
    ],
    "full_name": "Assistant User",
    "email": "assistant-user@elastic.co",
    "metadata": {},
    "enabled": true
}
```
</p>
</details>

Test that assistant is **NOT** available in UI via `Chat` buttons or
shortcut keys.

---

### Serverless Testing:

To test with the Assistant available, set `productTypes` to `complete`
in `config/serverless.security.yml`

```
xpack.securitySolutionServerless.productTypes:
  [
    { product_line: 'security', product_tier: 'complete' },
    { product_line: 'endpoint', product_tier: 'complete' },
  ]
  ```

otherwise to test without the Assistant, pick a different product type like `essentials`:

```
xpack.securitySolutionServerless.productTypes:
  [
    { product_line: 'security', product_tier: 'essentials' },
    { product_line: 'endpoint', product_tier: 'essentials' },
  ]
```

Then start Serverless Kibana: `yarn serverless-security`

---

### Checklist

Delete any items that are not applicable to this PR.

- [X] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

PhilippeOberti committed:
[Security Solution] expandable flyout - inverse Visualizations and Investigation order and expand Investigation by default (#163684)

wandergeek committed:
Add timeWindow type changes to slo docs (#163367)

mattkime committed:
[data views] swap_references api improvements (#163225)

## Summary

Some simple dev UX improvements to the swap_references data views api - 

```
POST /api/data_views/swap_references/_preview
{
     "fromId" : "abcd-efg",
     "toId" : "xyz-123"
}

returns 
{
  result: [{ id: "123", type: "visualization" }],
}
```


```
POST /api/data_views/swap_references
{
     "fromId" : "abcd-efg",
     "toId" : "xyz-123",
     "delete" : true // optional, removes data view which is no longer referenced
}

returns 
{
  result: [{ id: "123", type: "visualization" }],
  deleteStatus: {
    remainingRefs: 0,
    deletePerformed: true
}
```

Additional params - 

```
fromType: string - specify the saved object type. Default is `index-pattern` for data view
forId: string | string[] - limit the affected saved objects to one or more by id
forType: string - limit the affected saved objects by type
```

Improves upon https://github.com/elastic/kibana/pull/157665

Docs will be created in follow up PR

dhurley14 committed:
[Security Solution] [Detections] Fixes flakey exceptions read-only viewer cypress test (#162839)

## Summary

Ref: https://github.com/elastic/kibana/issues/162569

The test was trying to load the exceptions tab before the rule details
page loaded. Now we wait for the rule tab to load before continuing.
Something I was unaware of was that `cy.url()` will [automatically
retry](https://docs.cypress.io/api/commands/url#Assertions) until all
chained assertions have passed, which I think can be an easy way to fix
future flake issues where cypress tries to click on an element before
the new page loads.

davismcphee commented on src/plugins/discover/server/locator/searchsource_from_locator.ts:
```diff
@@ -147,7 +148,17 @@ export function searchSourceFromLocatorFactory(services: LocatorServicesDeps) {
     if (savedSearch.sort) {
-      const sort = getSortForSearchSource(savedSearch.sort as Array<[string, string]>, index);
+      const defaultSortingSetting = await services.uiSettings.get(SORT_DEFAULT_ORDER_SETTING);
+      const uiSettingsSyncReplacement = {
```
What's this code doing and does it change the behaviour of reporting at all?

jughosta replied:
@davismcphee It now reads `SORT_DEFAULT_ORDER_SETTING` setting. Before is was implicitly using `desc`.

jughosta replied:
Would you suggest to revert?

cc @kertal

davismcphee reviewed (commented):
This is looking really good! I didn't get a chance to do much testing yet but I went through the code and left some comments and questions. Will follow up with feedback from my testing next week.

kibanamachine committed:
[api-docs] 2023-08-12 Daily api_docs build (#163762)

Generated by
https://buildkite.com/elastic/kibana-api-docs-daily/builds/427

afharo committed:
[Telemetry] Use header-based versioned APIs instead of path-based (#159839)

kibanamachine committed:
[api-docs] 2023-08-13 Daily api_docs build (#163773)

Generated by
https://buildkite.com/elastic/kibana-api-docs-daily/builds/428

dgieselaar committed:
[Observability AI Assistant] Update README.md (#163769)

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Coen Warmer <coen.warmer@gmail.com>

kibanamachine committed:
[api-docs] 2023-08-14 Daily api_docs build (#163775)

Generated by
https://buildkite.com/elastic/kibana-api-docs-daily/builds/429

parkiino committed:
[Fleet][Agent Policy][Agent Tamper Protection] UI / API guard agent tamper protection only available if security defend integration present (#162196)

## Summary
UI
- [x] When there is no elastic defend integration present, the agent
tamper protection (`is_protected`) switch and instruction link are
disabled and there is an info tooltip explaining why the switch is
disabled

API
- [x] Requires the elastic defend integration to be present, in order to
set `is_protected` to true. Will allow the user to create the agent
policy and not throw an error, but will keep `is_protected` as false and
log a warning in the kibana server. In the next release, the response
will be modified to send back a 201 with the relevant messaging.
- [x] Sets `is_protected` to false when a user deletes the elastic
defend package policy

## Screenshots

### No Elastic Defend integration installed
<img width="970" alt="image"
src="https://github.com/elastic/kibana/assets/56409205/910be766-1a1e-4580-9ace-306089b4626d">

hop-dev committed:
[Fleet] Only enable secret storage once all fleet servers are above 8.10.0 (#163627)

## Summary

Closes #157456

Secret storage requires that fleet servers are 8.10.0 or above. 

This PR adds a backend check that all fleet servers are above 8.10.0
before enabling secrets storage. Once all fleet servers are above that
version, secrets are permanently enabled.

the fleet server check checks all agents in policies that contain the
fleet server package.

A flag on the`ingest_manager_settings` saved. object
`secret_storage_requirements_met` is used to make a note that the check
has previously passed, meaning we don't have to keep querying the agents
and policies.

Test scenarios (all covered by integration tests) : 

- given a deployment with no fleet servers connected, on creating a
package policy with secret variables, the values should be stored in
plain text not as a secret reference
- given a deployment with at least one fleet server that is below
8.10.0, on creating a package policy with secret variables, the values
should be stored in plain text not as a secret reference
- given a deployment where all fleet servers are 8.10.0 or above,
secrets should be stored as secret references and in the secrets index
- if a package policy was created before secrets were enabled, and since
its creation the fleet server versions pass the check, when updating
that policy, all secrets should move to being secret references.

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Julia Bardi <julia.bardi@elastic.co>
Co-authored-by: Julia Bardi <90178898+juliaElastic@users.noreply.github.com>

jughosta committed:
[Discover] Send a performance metric for fetchMore

jughosta commented:
Sorry all for the ping
